src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6266,13 +6335,117 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
             {
                 return this.ParseTupleType();
             }
+            else if (IsFunctionPointerStart())
+            {
+                return ParseFunctionPointerTypeSyntax();
+            }
             else
             {
                 var name = this.CreateMissingIdentifierName();
                 return this.AddError(name, ErrorCode.ERR_TypeExpected);
             }
         }
 
+#nullable enable
+        private FunctionPointerTypeSyntax ParseFunctionPointerTypeSyntax()
+        {
+            Debug.Assert(IsFunctionPointerStart());
+            var @delegate = EatToken(SyntaxKind.DelegateKeyword);
+            var asterisk = EatToken(SyntaxKind.AsteriskToken);
+            // Invalid calling conventions will be reported at type binding time
+            var callingConvention = IsPossibleCallingConvention() ? EatTokenAsKind(SyntaxKind.IdentifierToken, reportError: false) : null;
+
+            if (!IsPossibleFunctionPointerParameterListStart(CurrentToken))
+            {
+                var lessThanTokenError = SyntaxFactory.MissingToken(SyntaxKind.LessThanToken);
+                lessThanTokenError = WithAdditionalDiagnostics(lessThanTokenError, GetExpectedTokenError(SyntaxKind.LessThanToken, SyntaxKind.None));
+                var typesError = _pool.AllocateSeparated<FunctionPointerParameterOrReturnTypeSyntax>();
+                var identifierError = CreateMissingIdentifierName();
+                identifierError = AddError(identifierError, ErrorCode.ERR_TypeExpected);
+                var typeError = SyntaxFactory.FunctionPointerParameterOrReturnType(default, identifierError);
+                typesError.Add(typeError);
+                var greaterThanTokenError = SyntaxFactory.MissingToken(SyntaxKind.GreaterThanToken);
+                greaterThanTokenError = WithAdditionalDiagnostics(greaterThanTokenError, GetExpectedTokenError(SyntaxKind.GreaterThanToken, SyntaxKind.None));
+
+                var funcPtr = SyntaxFactory.FunctionPointerType(@delegate, asterisk, callingConvention, lessThanTokenError, typesError, greaterThanTokenError);
+                _pool.Free(typesError);
+                return funcPtr;
+            }
+
+            var lessThanToken = EatToken(SyntaxKind.LessThanToken);
+            var saveTerm = _termState;
+            _termState |= TerminatorState.IsEndOfFunctionPointerParameterList;
+            var types = _pool.AllocateSeparated<FunctionPointerParameterOrReturnTypeSyntax>();
+
+            try
+            {
+                while (true)
+                {
+                    var modifiers = _pool.Allocate<SyntaxToken>();
+                    try
+                    {
+                        // Modifiers will be checked for validity in later steps
+                        while (IsParameterModifier(CurrentToken.Kind) || CurrentToken.Kind == SyntaxKind.ReadOnlyKeyword)
```

Comment:
repeated with the scan code.  consider extracting into IsFunctionPointerModifier